balena-cli: init at 15.1.1

kalebpace opened the pull request:
###### Description of changes

This PR includes the [balena-cli](https://github.com/balena-io/balena-cli) tool.
"...It is a software tool available for Windows, macOS and Linux, used through a command prompt / terminal window. It can be used interactively or invoked in scripts. The balena CLI builds on the [balena API](https://www.balena.io/docs/reference/api/overview/) and the [balena SDK](https://www.balena.io/docs/reference/sdk/node-sdk/), and can also be directly imported in Node.js applications..."

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] x86_64-darwin (rosetta)
  - [ ] x86_64-windows
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

kalebpace committed:
balena-cli: init at 15.1.1

balena-cli: init at v14.3.1

balena-cli: update to 15.1.1

AndersonTorres commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,66 @@
+
+      license = licenses.asl20;
+      maintainers = [ maintainers.kalebpace ];
+      platforms = platforms.all;
```
```suggestion
      platforms = with platforms; linux ++ darwin;
```

Indeed we should restrict even more, but I don't remember the exact expression for this.

kalebpace replied:
Thanks for catching this

AndersonTorres commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,66 @@
+      license = licenses.asl20;
+      maintainers = [ maintainers.kalebpace ];
+      platforms = platforms.all;
+    };
```
```suggestion
      sourceProvenance = with sourceTypes; [ binaryNativeCode ];
    };
```

kalebpace commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,66 @@
+      mkdir -p $out/bin
+      cp -r ./* $out/
+
+      ln -s $out/balena $out/bin/balena
```
Unfortunately, balena-cli does not bundle itself into a single excutable, and relies on modules/directories within the directory the executable runs from. So I just linked it. 

Btw, thank you for taking time to review!

AndersonTorres reviewed (changes requested):
Squash the commits into only one, with the message `balena-cli: init at 14.3.1`

kalebpace commented:
> @AndersonTorres @kalebpace Any updates on this? Once updated to `v15.1.0` it should be good to merge, right?

Just updated the package to `v15.1.1`. Not sure what else is needed to be merged though, this is my first nix package submission.

AndersonTorres commented:
github continuously complained about "unable to merge", "conflicting lines" and such.

I haven't touched this review since then.

kalebpace commented:
Heard! Not pointing fingers. Was unaware of conflicts since last time I squashed commits. Fixed some that came up this time around. Will wait for OfBorg to complete and see where we are at. 

Happy to modify as needed, address nits, etc. Just lmk.

superherointj commented:
@kalebpace 

> pkgs/tools/admin/balena-cli/default.nix:
>	15: Trailing whitespace

Please fix.

superherointj commented on maintainers/maintainer-list.nix:
```diff
@@ -7228,6 +7228,13 @@
     githubId = 87115;
     name = "Wael Nasreddine";
   };
+  kalebpace = {
+    email = "kaleb.pace@pm.me";
+    matrix = "@kalebpace:matrix.org";
+    github = "kalebpace";
+    githubId = 5586615;
+    name = "Kaleb Pace";
+  };
```
```suggestion
```

superherointj replied:
Needs a rebase.

superherointj commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,68 @@
+, stdenv
+, fetchurl
+, unzip
+, installShellFiles
```
```suggestion
, installShellFiles
```
It's declared but not used?

superherointj commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,68 @@
+  pname = "balena-cli";
+  version = "15.1.1";
+
+  src = fetchurl {
```
```suggestion
  src = fetchurl {
```
Wouldn't fetchzip work?

> nativeBuildInputs = [ unzip ];

Is it declared because of that?

kalebpace replied:
Yep. What's a good way to compute the SHA for use with fetchzip? I just tried and due to it uncompressing before comparing SHA's, the archive SHA is no longer valid. If I generate a SHA for the uncompressed directory, won't it differ between platforms/filesystems?

superherointj commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,68 @@
+
+  meta = with lib; {
+    mainProgram = "balena";
+    description = "The official balena Command Line Interface";
```
```suggestion
    description = "A command line interface for balenaCloud or openBalena";
```

superherointj commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,68 @@
+  '';
+
+  meta = with lib; {
+    mainProgram = "balena";
```
```suggestion
    mainProgram = "balena";
```
Please move mainProgram to bottom of meta. Having description at top is best. There is order to meta attributes. I know it may sound odd. But there are reasons for this. For example, `broken` is always last at meta so it doesn't create git conflicts when marking/unmarking broken.

superherointj commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,68 @@
+      It can be used interactively or invoked in scripts. The balena CLI builds on the balena API
+      and the balena SDK, and can also be directly imported in Node.js applications.
+    '';
+
```
```suggestion
```

superherointj commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,68 @@
+
+    homepage = "https://github.com/balena-io/balena-cli";
+    changelog = "https://github.com/balena-io/balena-cli/blob/v${version}/CHANGELOG.md";
+
```
```suggestion
```
Removing empty line.

AndersonTorres commented on pkgs/tools/admin/balena-cli/default.nix:
```diff
@@ -0,0 +1,61 @@
+    changelog = "https://github.com/balena-io/balena-cli/blob/v${version}/CHANGELOG.md";
+    license = licenses.asl20;
+    maintainers = [ maintainers.kalebpace ];
+    platforms = with platforms; linux ++ darwin ++ cygwin ++ windows;
```
It doesn't work.
[As I have documented another day](https://github.com/atorres1985-contrib/nixpkgs/commit/451dfa6d84586b5cb83e6a7254689dff66b4523a), `with` is not distributive in relation to `++`: 

https://github.com/pikajude/nixpkgs/commit/eea78e554e03906524f87406bb3a17d5c5264029


```suggestion
    platforms = platforms.linux ++ platforms.darwin ++ platforms.cygwin ++ platforms.windows;
```

AndersonTorres reviewed (changes requested):
Besides LGTM

superherointj commented:
@kalebpace 

![image](https://user-images.githubusercontent.com/5861043/226105860-9ab94a06-6f74-47fb-aac8-962d72caf957.png)
Is it working as intended?

I was trying to verify is version is correct. Can you check that too?

mguentner commented:
I can confirm that it is not working on NixOS as the binary (balena) is not patched.

I tried to add `autoPatchelfHook` and `stdenv.cc.cc.lib` to the `nativeBuildInputs` like this:

```
  nativeBuildInputs = [ unzip stdenv.cc.cc.lib autoPatchelfHook ];
```
which allows the derivation to build + patch the binaries but leads to this awesome error:

```
❯ balena
pkg/prelude/bootstrap.js:1
,"index.js"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["extend-shallow"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["package.json","index.js"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["package.json","index.js","node_modules"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["package.json","index.js"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["package.json","index.js"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["package.json","index.js"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["define-property"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["package.json","index.js"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["package.json","index.js"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["package.json","index.js"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["is-extendable"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["extend-shallow"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["extend-shallow"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["package.json","index.js"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["extend-shallow"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["has-value","has-values"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["define-property"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["is-number","kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["isobject"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["define-property","kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}["kind-of"]{"mode":16877,"size":0,"isFileValue":false,"isDirectoryValue":true,"isSocketValue":false,"isSymbolicLinkValue":false}(function(process, require, console, EXECPATH_FD, PAYLOAD_POSITION, PAYLOAD_SIZE) { return (function (REQUIRE_COMMON, VIRTUAL_FILESYSTEM, DEFAULT_ENTRYPOINT, SYMLINKS, DICT, DOCOMPRESS) { 
^

SyntaxError: Unexpected token ','
    at new Script (vm.js:103:7)
    at readPrelude (internal/bootstrap/pkg.js:29:13)
    at internal/bootstrap/pkg.js:36:18
    at internal/bootstrap/pkg.js:43:4
    at internal/bootstrap/pkg.js:44:2

```

I resorted to run the binary with [nix-ld](https://github.com/Mic92/nix-ld) in the meantime.

superherointj commented:
@mguentner Thanks for reporting. Can you open an issue reporting breakage?
Feel free to add people involved here there.

kalebpace commented:
Interesting, it was working for me before merging. But I can confirm that what is in master is no longer functional.

I just removed the `pre/postInstallHook` which fixes it for me again. I'll also test this on my x86 NixOS machine for mguentner and submit a PR fixing both issues if I can, and link it here.
